src/main/java/at/ac/tuwien/kr/alpha/grounder/instantiation/AbstractLiteralInstantiationStrategy.java

```diff
@@ -0,0 +1,137 @@
+	 * literals is determined using the abstract method {@link AbstractLiteralInstantiationStrategy#getAssignmentStatusForAtom(Atom)}
+	 */
+	@Override
+	public final AssignmentStatus getTruthForGroundLiteral(Literal groundLiteral) {
+		if (groundLiteral.isNegated()) {
+			return this.getAssignmentStatusForNegatedGroundLiteral(groundLiteral);
+		}
+		return this.getAssignmentStatusForAtom(groundLiteral.getAtom());
+	}
+
+	/**
+	 * See {@link LiteralInstantiationStrategy#getAcceptedSubstitutions(Literal, Substitution)}.
+	 * 
+	 * A very general implementation of the basic steps needed to obtain ground substitutions for a positive literal.
+	 * Potentially valid ground instances are obtained using {@link AbstractLiteralInstantiationStrategy#computeCandidateInstances(Atom)}, then
+	 * checked, such that each candidate instance has unifies with the given partial substitution and has a "valid" {@link AssignmentStatus},
+	 * where "validity" of an {@link AssignmentStatus} is determined using the abstract method
+	 * {@link AbstractLiteralInstantiationStrategy#assignmentStatusAccepted(AssignmentStatus)}.
+	 */
+	@Override
+	public final List<ImmutablePair<Substitution, AssignmentStatus>> getAcceptedSubstitutions(Literal lit, Substitution partialSubstitution) {
+		Atom atom = lit.getAtom();
+		Iterable<Instance> groundInstances = this.computeCandidateInstances(atom);
+		return this.buildSubstitutionsFromInstances(atom, groundInstances, partialSubstitution);
+	}
+
+	/**
+	 * Computes instances that are potentially valid ground instances of the given partially-ground atom.
+	 * 
+	 * A candidate instance is a ground instance of the same predicate where all terms that are ground in <code>partiallyGroundAtom</code> have
+	 * the same values in the candidate.
+	 * 
+	 * @param partiallyGroundAtom a partially ground atom for which to find fitting ground instances
+	 * @return a list of candidate instances
+	 */
+	protected abstract Iterable<Instance> computeCandidateInstances(Atom partiallyGroundAtom);
+
+	/**
+	 * Based on a list of candidate instances (see {@link AbstractLiteralInstantiationStrategy#computeCandidateInstances(Atom)}), create a list
+	 * of substitutions and assignment statuses such that each substitution represents a valid (according to the implementation-specific
+	 * definition of this instantiation strategy) ground instance of <code>atomToSubstitute</code>
+	 * 
+	 * @param atomToSubstitute
+	 * @param candidateInstances
+	 * @param partialSubstitution
+	 * @return
+	 */
+	protected final List<ImmutablePair<Substitution, AssignmentStatus>> buildSubstitutionsFromInstances(Atom atomToSubstitute,
+			Iterable<Instance> candidateInstances, Substitution partialSubstitution) {
+		List<ImmutablePair<Substitution, AssignmentStatus>> retVal = new ArrayList<>();
+		// Filter for only instances unifying with partialSubsitution,
+		// i.e. "where all joins work out".
+		Substitution currentInstanceSubstitution;
+		Atom atomForCurrentInstance;
```

Comment:
Not really other than the fact that it seems the more natural way to me rather than re-defining them inside the loop all the time.

Reply:
In general, I would rather tend to defining those variables inside the loop, since they clearly are local variables only important or used inside the loop. Furthermore, if defined inside the loop, it is immediately clear that the value does not depend other iterations of the loop, as it can only depend on the current iteration.
As regards efficiency, I assume that the compiler/JVM is turning the variable into one on the function stack anyway (or changes it into whatever it deems most efficient, e.g. a CPU register).
Note: just commenting to express my view on that point, no code change is required as this is no big thing.